File: check_history.py
#!/usr/bin/env python3
"""Check invoice registration status for all T-numbers in history.csv"""

import csv
import re
import subprocess
import sys
import time

CSV_PATH = "history.csv"
UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
BASE_URL = "https://www.invoice-kohyo.nta.go.jp/regno-search/detail"
TNUM_COL = 4  # 0-indexed: column E = "インボイス番号"

def check_tnum(digits, ua=UA):
    """Check if a T-number is registered. Returns True/False/None."""
    url = f"{BASE_URL}?selRegNo={digits}"
    try:
        result = subprocess.run(
            ["curl", "-s", "-A", ua, "--connect-timeout", "10", "--max-time", "15", url],
            capture_output=True, text=True, timeout=20
        )
        html = result.stdout
        if "検索対象の登録番号は存在しません" in html:
            return False
        if "登録番号" in html:
            return True
        return None
    except Exception as e:
        print(f"  ERROR: {e}", file=sys.stderr)
        return None

def normalize_tnum(raw):
    """Normalize T-number. Returns (digits, is_valid)."""
    raw = raw.strip()
    if not raw or raw == "なし":
        return None, False
    if raw.startswith("T"):
        digits = raw[1:]
    else:
        digits = raw
    if len(digits) == 13 and digits.isdigit():
        return digits, True
    return None, False

def main():
    # Read CSV
    with open(CSV_PATH, "r", encoding="utf-8") as f:
        reader = csv.reader(f)
        rows = list(reader)

    if not rows:
        print("Empty CSV")
        return

    header = rows[0]
    data = rows[1:]
    print(f"Loaded {len(data)} data rows")

    # Collect unique T-numbers
    unique_digits = {}
    for row in data:
        if len(row) > TNUM_COL:
            digits, valid = normalize_tnum(row[TNUM_COL])
            if valid and digits not in unique_digits:
                unique_digits[digits] = None  # placeholder

    print(f"Found {len(unique_digits)} unique T-numbers to check")

    # Check each unique T-number
    for i, digits in enumerate(unique_digits.keys()):
        registered = check_tnum(digits)
        unique_digits[digits] = registered
        status = "登録済" if registered else ("未登録" if registered is False else "確認不可")
        print(f"  [{i+1}/{len(unique_digits)}] T{digits} => {status}")
        time.sleep(0.1)  # polite delay

    # Update rows
    header.append("登録状況")
    for row in data:
        if len(row) > TNUM_COL:
            digits, valid = normalize_tnum(row[TNUM_COL])
            if valid:
                reg = unique_digits.get(digits)
                if reg is True:
                    row.append("登録済")
                elif reg is False:
                    row.append("未登録")
                else:
                    row.append("確認不可")
            else:
                row.append("番号なし")
        else:
            row.append("番号なし")

    # Write back
    with open(CSV_PATH, "w", encoding="utf-8", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(header)
        writer.writerows(data)

    # Summary
    registered = sum(1 for r in data if r[-1] == "登録済")
    unregistered = sum(1 for r in data if r[-1] == "未登録")
    no_number = sum(1 for r in data if r[-1] == "番号なし")
    unknown = sum(1 for r in data if r[-1] == "確認不可")
    print(f"\n=== Summary ===")
    print(f"  登録済: {registered}")
    print(f"  未登録: {unregistered}")
    print(f"  番号なし: {no_number}")
    print(f"  確認不可: {unknown}")
    print(f"  合計: {len(data)} rows")
    print(f"Updated {CSV_PATH}")

if __name__ == "__main__":
    main()

File: check_history.mjs
import { readFileSync, writeFileSync } from 'fs';
import { execSync } from 'child_process';

const CSV_PATH = 'history.csv';
const UA = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36';
const BASE_URL = 'https://www.invoice-kohyo.nta.go.jp/regno-search/detail';
const TNUM_COL = 4;

function parseCSVLine(line) {
    const fields = [];
    let current = '';
    let inQuotes = false;
    for (const ch of line) {
        if (ch === '"') { inQuotes = !inQuotes; }
        else if (ch === ',' && !inQuotes) { fields.push(current); current = ''; }
        else { current += ch; }
    }
    fields.push(current);
    return fields;
}

function normalizeTnum(raw) {
    raw = raw.trim();
    if (!raw || raw === 'なし') return null;
    let digits = raw.startsWith('T') ? raw.slice(1) : raw;
    if (digits.length === 13 && /^\d{13}$/.test(digits)) return digits;
    return null;
}

function checkTnum(digits) {
    const url = `${BASE_URL}?selRegNo=${digits}`;
    try {
        const html = execSync(
            `curl -s -A "${UA}" --connect-timeout 10 --max-time 15 "${url}"`,
            { encoding: 'utf-8', timeout: 20000 }
        );
        if (html.includes('検索対象の登録番号は存在しません')) return false;
        if (html.includes('登録番号')) return true;
        return null;
    } catch (e) {
        console.error(`  ERROR checking T${digits}: ${e.message}`);
        return null;
    }
}

// Read CSV
const content = readFileSync(CSV_PATH, 'utf-8');
const lines = content.split('\n').filter(l => l.trim());
const header = parseCSVLine(lines[0]);
const dataRows = lines.slice(1).map(l => parseCSVLine(l));

console.log(`Loaded ${dataRows.length} data rows`);

// Collect unique T-numbers
const uniqueDigits = new Map();
for (const row of dataRows) {
    if (row.length > TNUM_COL) {
        const digits = normalizeTnum(row[TNUM_COL]);
        if (digits && !uniqueDigits.has(digits)) {
            uniqueDigits.set(digits, null);
        }
    }
}

console.log(`Found ${uniqueDigits.size} unique T-numbers to check`);

// Check each unique T-number
let idx = 0;
for (const [digits] of uniqueDigits) {
    idx++;
    const registered = checkTnum(digits);
    uniqueDigits.set(digits, registered);
    const status = registered === true ? '登録済' : (registered === false ? '未登録' : '確認不可');
    console.log(`  [${idx}/${uniqueDigits.size}] T${digits} => ${status}`);
}

// Update rows
header.push('登録状況');
for (const row of dataRows) {
    if (row.length > TNUM_COL) {
        const digits = normalizeTnum(row[TNUM_COL]);
        if (digits) {
            const reg = uniqueDigits.get(digits);
            row.push(reg === true ? '登録済' : (reg === false ? '未登録' : '確認不可'));
        } else {
            row.push('番号なし');
        }
    } else {
        row.push('番号なし');
    }
}

// Write CSV back
const outputLines = [header.join(',')];
for (const row of dataRows) {
    // Re-quote fields containing commas
    const quoted = row.map(f => f.includes(',') ? `"${f}"` : f);
    outputLines.push(quoted.join(','));
}
writeFileSync(CSV_PATH, outputLines.join('\n') + '\n', 'utf-8');

// Summary
const counts = { '登録済': 0, '未登録': 0, '番号なし': 0, '確認不可': 0 };
for (const row of dataRows) {
    const s = row[row.length - 1];
    if (counts[s] !== undefined) counts[s]++;
}
console.log(`\n=== Summary ===`);
console.log(`  登録済: ${counts['登録済']}`);
console.log(`  未登録: ${counts['未登録']}`);
console.log(`  番号なし: ${counts['番号なし']}`);
console.log(`  確認不可: ${counts['確認不可']}`);
console.log(`  合計: ${dataRows.length} rows`);
console.log(`Updated ${CSV_PATH}`);
